### PR TITLE
SCP-4986 Integration test scenarios for Intersections

### PR DESCRIPTION
**Pull request by jhbertra**

### CODE REVIEW

**Review by bwbush — APPROVED** [review]
Nice framework!

I'm concerned that minor, occasional lack of synchronization on the backend will cause intermittent failures. Should we add "knobs" like for a delay after a confirmation, etc.

Also, I suggest we fix the CI failures on `main` before merging this.

**Comment by jhbertra** [comment]
> Nice framework!
> 
> I'm concerned that minor, occasional lack of synchronization on the backend will cause intermittent failures. Should we add "knobs" like for a delay after a confirmation, etc.

Not preemptively - I'd want to wait for there to be an actual failure to diagnose first

> Also, I suggest we fix the CI failures on `main` before merging this.

I just pushed a couple of commits to one of your branches that should address this.

File: marlowe-integration-tests/test/Language/Marlowe/Runtime/Integration/StandardContract.hs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  , gimmeTheMoneyChosen :: InputsApplied BabbageEra v
+  , choiceBlock :: BlockHeader
+  }
+
+data StandardContractNotified v = StandardContractNotified
+  { makeReturnDeposit :: Integration (StandardContractClosed v)
+  , notified :: InputsApplied BabbageEra v
+  , notifiedBlock :: BlockHeader
+  }
+
```
[the remark this answers is not in the document]
Comment (jhbertra) [reply marlowe-integration-tests/test/Language/Marlowe/Runtime/Integration/StandardContract.hs:52]: I think that's better to have a test separately to cover - it requires a bit more attention to time intervals and waiting for the timeout period to elapse.

**Review by ladamesny — APPROVED** [review]

**Comment by jhbertra** [comment]
> > Nice framework!
> > I'm concerned that minor, occasional lack of synchronization on the backend will cause intermittent failures. Should we add "knobs" like for a delay after a confirmation, etc.
> 
> Not preemptively - I'd want to wait for there to be an actual failure to diagnose first

Well here's an example of a concrete failure related to this :) https://github.com/input-output-hk/marlowe-cardano/actions/runs/4118311156/jobs/7110668432

What I believe is happening here is that we are submitting the transaction, it is getting added to the chain, and `marlowe-tx` is returning the block header. Then before we call `Poll` / `RequestNext` in the sync protocol and get the new tx and block header, it has been rolled back and reapplied at a later slot no, so the `expectedBlock` does not match the `actualBlock`.

I see two options here:

1. Remove the assertions for the block headers, since we cannot be certain they will actually match, though this is not a bug.
2. We add a parameter (or "knob" if you wish) to `marlowe-tx` which requires it to wait X blocks after it receives the confirmation for the submitted transaction before it returns the block header, and we turn up the value a bit in the tests.

**Pull request merged**